### DISCUSSION TITLE
Recommended way to build MicroPython as a library?

### DISCUSSION THREAD

**Discussion by bogdanm** [micropython/micropython#10906] [Core Development]
Hi, I am looking at using MicroPython in a project, but not in its standalone variant. This means that the project has its own source tree (including its own `main` function) and will initialize and start MicroPython as needed, so I'm looking for a way to build MicroPython as a library that I can link with my own code. I managed to do that, but I'm wondering if there's a simpler (or "more standard") way to do this that I'm missing. My application uses Zephyr, so I'm using the Zephyr port or MicroPython as follows:

1. I added these lines to my top level `CMakeLists.txt`:

```cmake
# Set the MicroPython port configuration options before including it
set(CONFIG_MICROPY_CONFIGFILE "mpconfigport.h")
set(CONFIG_MICROPY_HEAP_SIZE 16384)
add_subdirectory(${CMAKE_SOURCE_DIR}/micropython/ports/zephyr)
add_dependencies(app micropython)
```

2.  I modified `ports/zephyr/CMakeLists.txt` as follows:

```diff
--- a/ports/zephyr/CMakeLists.txt
+++ b/ports/zephyr/CMakeLists.txt
@@ -101,6 +100,8 @@ zephyr_library_include_directories(
     ${MICROPY_INC_CORE}
     ${MICROPY_PORT_DIR}
     ${CMAKE_CURRENT_BINARY_DIR}
+    ${CMAKE_CURRENT_SOURCE_DIR}
+    ${CMAKE_BINARY_DIR}
 )

 zephyr_library_compile_options(
@@ -123,9 +124,9 @@ add_dependencies(${MICROPY_TARGET} zephyr_generated_headers)

 include(${MICROPY_DIR}/py/mkrules.cmake)

-target_sources(app PRIVATE
-    src/zephyr_start.c
-    src/zephyr_getchar.c
-)
+#target_sources(app PRIVATE
+#    src/zephyr_start.c
+#    src/zephyr_getchar.c
+#)
```

In other words:

- I removed the parts related to generating an executable image (`target_sources`).
- I added `CMAKE_BINARY_DIR` to the include path, because some tools from the build system (like `genhdr`) output into `CMAKE_BINARY_DIR` instead of `CMAKE_CURRENT_BINARY_DIR` (which is relevant when `add_subdirectory` is used).

While this seems to do the trick, I'm wondering if there's a way to achieve this that doesn't require me to patch `ports/zephyr/CMakeLists.txt`.
